torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,212 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+#ifdef ncclGroupStart
+#define NCCL_VERSION 2
+#endif
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, std::unordered_map<int, ncclComm_t> > _communicators;
+std::vector<ncclComm_t *> _all_communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto comm: _all_communicators) {
+    ncclCommDestroy(*comm);
+  }
+  Py_RETURN_NONE;
+}
+
+static std::unordered_map<int, ncclComm_t> _get_communicator(const int *devs, int ndevices) {
+  std::stringstream hash;
+  for (int i = 0; i < ndevices; i++) {
+    hash << devs[i] << ",";
+  }
+  if (_communicators.find(hash.str()) == _communicators.end()) {
+    ncclComm_t *comms = (ncclComm_t*) malloc(sizeof(ncclComm_t) * ndevices);
+    assert(comms);
+    CHECK(ncclCommInitAll(comms, ndevices, devs));
+    _all_communicators.push_back(comms);
+    for (int i = 0; i < ndevices; i++) {
+      _communicators[hash.str()][devs[i]] = comms[i];
+    }
+  }
+  return _communicators[hash.str()];
+}
+
+static void _check_inputs(THPObjectPtr &inputs, THPObjectPtr &outputs, int size_multiplier) {
+  if (!PySequence_Check(inputs.get()) || !PySequence_Check(outputs.get())) {
+    throw std::runtime_error("inputs and outputs have to be sequences of Tensors");
+  }
+
+  // len(inputs) == len(outputs)
+  Py_ssize_t len = PySequence_Fast_GET_SIZE(inputs.get());
+
+  if (len <= 0) {
+    throw std::runtime_error("input sequence can't be empty");
+  }
```

Comment:
`THPUtils_assert` is a shorter version of this. Use that for all checks

Reply:
this is only true in this function, because it is directly called from python. in child functions you cant use it. to keep things consistent in this file, i used std::runtime_error everywhere.